testdata/schema.go

```diff
@@ -0,0 +1,232 @@
+	newFields := make([]arrow.Field, len(fields))
+	copy(newFields, fields)
+	for _, d := range dt {
+		for i := 0; i < len(newFields); i++ {
+			if arrow.TypeEqual(newFields[i].Type, d) {
+				newFields = append(newFields[:i], newFields[i+1:]...)
+				i--
+			}
+		}
+	}
+	return newFields
+}
+
+// ListOfFields returns a list of fields that are lists of the given fields.
+func ListOfFields(baseFields []arrow.Field) []arrow.Field {
+	fields := make([]arrow.Field, len(baseFields))
+	for i := 0; i < len(baseFields); i++ {
+		fields[i] = arrow.Field{Name: baseFields[i].Name + "_list", Type: arrow.ListOf(baseFields[i].Type), Nullable: true}
+	}
+	return fields
+}
+
+// MapOfFields returns a list of fields that are maps of the given fields.
+func MapOfFields(baseFields []arrow.Field) []arrow.Field {
+	fields := make([]arrow.Field, len(baseFields))
+	for i := 0; i < len(baseFields); i++ {
+		fields[i] = arrow.Field{Name: baseFields[i].Name + "_map", Type: arrow.MapOf(baseFields[i].Type, baseFields[i].Type), Nullable: true}
+	}
+	return fields
+}
+
+// TestSourceOptions controls which types are included by TestSourceFields.
+type TestSourceOptions struct {
+	IncludeLists      bool // lists of all primitive types. Lists that were supported by CQTypes are always included.
+	IncludeTimestamps bool // timestamp types. Microsecond timestamp is always be included, regardless of this setting.
+	IncludeDates      bool
+	IncludeMaps       bool
+	IncludeStructs    bool
+	IncludeIntervals  bool
+	IncludeDurations  bool
+	IncludeTimes      bool // time of day types
+	IncludeLargeTypes bool // e.g. large binary, large string
+}
+
+// TestSourceFields returns fields for all Arrow types and composites thereof. TestSourceOptions controls
+// which types are included.
+func TestSourceFields(opts TestSourceOptions) []arrow.Field {
+	// cq fields
+	var cqFields []arrow.Field
+	cqIDMetadata := arrow.NewMetadata([]string{schema.MetadataUnique}, []string{"true"})
+	cqFields = append(cqFields, arrow.Field{Name: schema.CqIDColumn.Name, Type: types.NewUUIDType(), Nullable: false, Metadata: cqIDMetadata})
+	cqFields = append(cqFields, arrow.Field{Name: schema.CqParentIDColumn.Name, Type: types.NewUUIDType(), Nullable: false})
+
+	var basicFields []arrow.Field
+	basicFields = append(basicFields, PrimitiveFields()...)
+	basicFields = append(basicFields, BinaryFields()...)
+	basicFields = append(basicFields, FixedWidthFields()...)
+
+	// add extensions
+	basicFields = append(basicFields, arrow.Field{Name: "uuid", Type: types.NewUUIDType(), Nullable: true})
+	basicFields = append(basicFields, arrow.Field{Name: "inet", Type: types.NewInetType(), Nullable: true})
+	basicFields = append(basicFields, arrow.Field{Name: "mac", Type: types.NewMacType(), Nullable: true})
+
+	// sort and remove duplicates (e.g. date32 and date64 appear twice)
+	basicFields = sortAndRemoveDuplicates(basicFields)
+
+	// we don't support float16 right now
```

Comment:
We can, though?
It should map.to float32 in the most cases

Reply:
We can, but that doesn't mean we should. As I said above, supporting new types comes with a cost of code that we need to maintain in all destinations right now. Until that changes, we should think carefully about what we add. In this case, I think we shouldn't support it:

 1. Go doesn't support float16, so no struct conversion will require it
 2. Postgres doesn't have float16, so the pg source plugin won't require it
 3. Even the official Arrow CSV and Parquet implementations don't support float16, so we would have to go contribute that upstream to support this